src/diagnostic-channel-publishers/src/pg.pub.ts

```diff
@@ -150,11 +150,163 @@ function postgres6PatchFunction(originalPg, originalPgPath) {
+    const originalClientQuery = originalPg.Client.prototype.query;
+    const diagnosticOriginalFunc = "__diagnosticOriginalFunc";
+
+    // wherever the callback is passed, find it, save it, and remove it from the call
+    // to the the original .query() function
+    originalPg.Client.prototype.query = function query(config, values, callback) {
+        let callbackProvided: boolean = !!callback; // Starting in pg@7.x+, Promise is returned only if !callbackProvided
+        const data: IPostgresData = {
+            query: {},
+            database: {
+                host: this.connectionParameters.host,
+                port: this.connectionParameters.port,
+            },
+            result: null,
+            error: null,
+            duration: 0,
+        };
+        const start = process.hrtime();
+        let queryResult;
+
+        function patchCallback(cb?: PostgresCallback): PostgresCallback {
+            if (cb && cb[diagnosticOriginalFunc]) {
+                cb = cb[diagnosticOriginalFunc];
+            }
+
+            const trackingCallback = channel.bindToContext(function(err: Error, res: IPostgresResult): any {
+                const end = process.hrtime(start);
+                data.result = res && { rowCount: res.rowCount, command: res.command };
+                data.error = err;
+                data.duration = Math.ceil((end[0] * 1e3) + (end[1] / 1e6));
+                channel.publish("postgres", data);
+
+                // emulate weird internal behavior in pg@6
+                // on success, the callback is called *before* query events are emitted
+                // on failure, the callback is called *instead of* the query emitting events
+                // with no events, that means no promises (since the promise is resolved/rejected in an event handler)
+                // since we are always inserting ourselves as a callback, we have to restore the original
+                // behavior if the user didn't provide one themselves
+                if (err) {
+                    if (cb) {
+                        return cb.apply(this, arguments);
+                    } else if (queryResult && queryResult instanceof EventEmitter) {
+                        queryResult.emit("error", err);
+                    }
+                } else if (cb) {
+                    cb.apply(this, arguments);
+                }
+            });
+
+            try {
+                Object.defineProperty(trackingCallback, diagnosticOriginalFunc, { value: cb });
+                return trackingCallback;
+            } catch (e) {
+                // this should never happen, but bailout in case it does
+                return cb;
+            }
+        }
+
+        // this function takes too many variations of arguments.
+        // this patches any provided callback or creates a new callback if one wasn't provided.
+        // since the callback is always called (if provided) in addition to always having a Promisified
```

Comment:
Is this comment true? Seems to not match description about callback behavior in pg7 above. Does this make any code here broken?

Reply:
It is not true anymore. Refactored to make it more correct